### PR TITLE
KAFKA-5362: Add EOS system tests for Streams API

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @guozhangwang @bbejeck @dguy

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+                committedOffsets.put(tp, offset);
+            }
+        }
+
+        return committedOffsets;
+    }
+
+    private static Map<String, Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>>> getOutputRecords(final KafkaConsumer<byte[], byte[]> consumer,
+                                                                                                           final Map<TopicPartition, Long> committedOffsets,
+                                                                                                           final List<TopicPartition> partitions) {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java]: partitions is not used

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+        final Map<String, Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>>> recordPerTopicPerPartition = new HashMap<>();
+
+        long maxWaitTime = System.currentTimeMillis() + 30000;
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java]: extract `30000` to a constant maybe?

**Review by dguy — COMMENTED** [review]
I'm wondering if we can simplify this test to only do a single operation, i.e., sum? Do we really need to have four different operations to prove that it works? The test seems complicated by the fact that it is trying to verify the various different results and, to me at least, it is adding a bunch of noise that i'm struggling to see through

**Comment by mjsax** [comment]
I designed the test similar to SmokeTest that does even more stuff... I still think, it makes sense to test simple "copy" and an "aggregation". I also want to make the test more complex and add a repartitioning step for one of the aggregates -- but it did not work yet, so I just removed the repartitioning...

Maybe we should split this into individual test though. WDYT?

**Comment by dguy** [comment]
I guess it becomes a trade off between system test runtime and code complexity. My preference would be to have simple tests that are easy to follow. If time becomes a problem then merge them.

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestClient.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                                                   final String kafka) {
+        final Properties props = new Properties();
+        props.put(StreamsConfig.APPLICATION_ID_CONFIG, APP_ID);
+        props.put(StreamsConfig.STATE_DIR_CONFIG, stateDir.toString());
+        props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        props.put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, 2);
+        props.put(StreamsConfig.NUM_STANDBY_REPLICAS_CONFIG, 2);
+        props.put(StreamsConfig.REPLICATION_FACTOR_CONFIG, 3);
+        props.put(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, StreamsConfig.EXACTLY_ONCE);
+        props.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/tests/EosTestClient.java:94]: Why do we want to disable caching? I think it is valuable to have at least one test case with caching turned on and see if EOS is still guaranteed?

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+                isRunning = false;
+            }
+        });
+
+        final Properties producerProps = new Properties();
+        producerProps.put(ProducerConfig.CLIENT_ID_CONFIG, "SmokeTest");
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class);
+        producerProps.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java:75]: Do you need to set the txnId also?

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+            || receivedRecords == null
+            || supersetExpectedRecords.keySet().size() < committedOffsets.keySet().size()
+            || receivedRecords.keySet().size() < committedOffsets.keySet().size()) {
+
+            return false;
+        }
+
+        for (final Map.Entry<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> partitionRecords : receivedRecords.entrySet()) {
+            final Long committed = committedOffsets.get(partitionRecords.getKey());
+            if (partitionRecords.getValue().size() < (committed == null ? 0 : committed)) {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java]: Is this okay? If committed == null it means no offsets get committed for this partition in `getCommittedOffsets`, and we should check that we do not get any data in `partitionRecords ` but currently this check will always pass even if `partitionRecords.getValue().size() > 0`.

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+                = getOutputRecords(consumer, committedOffsets, partitions);
+
+            truncate("data", recordPerTopicPerPartition, committedOffsets);
+
+            verifyMin(recordPerTopicPerPartition.get("data"), recordPerTopicPerPartition.get("min"));
+            verifySum(recordPerTopicPerPartition.get("data"), recordPerTopicPerPartition.get("sum"));
+
+            verifyAllTransactionFinished(consumer, kafka);
+
+            // no not modify: required test output
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java]: `no` -> `do`?

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+
+    private static void truncate(final String topic,
+                                 final Map<String, Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>>> recordPerTopicPerPartition,
+                                 final Map<TopicPartition, Long> committedOffsets) {
+        final Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> topicRecords = recordPerTopicPerPartition.get(topic);
+        final Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> truncatedTopicRecords = recordPerTopicPerPartition.get(topic);
+
+        for (final Map.Entry<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> partitionRecords : topicRecords.entrySet()) {
+            final TopicPartition tp = partitionRecords.getKey();
+            truncatedTopicRecords.put(tp, partitionRecords.getValue().subList(0, committedOffsets.get(tp).intValue()));
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java]: Similarly, `committedOffsets.get(tp).intValue()` could throw NPE?

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestClient.java (hunk 3 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/tests/EosTestClient.java:94]: With caching, we cannot know the result (ie, not each individual record that get's written to the output), as we don't know when we will flush. For this test, I want to check all intermediate records, too, so I need to disable caching.

But we can add a second test with caching and just make sure we don't over count for example and ignore all intermediate results and just check the latest result per key.

File: streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java (hunk 4 above)
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java:75]: We write non-transactional data. Actually, we might not even need idempotent producer as we assume that there is no error writing data for this test cases.

**Comment by mjsax** [comment]
Updated. 3 consecutive local runs passed.

**Comment by mjsax** [comment]
Triggered branch builder at: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/306/

**Review by bbejeck — APPROVED** [review]
LGTM

**Comment by mjsax** [comment]
Branch builder failed... Investigating.

**Comment by mjsax** [comment]
Branch builder passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/311/

**Comment by guozhangwang** [comment]
Merged to trunk and 0.11.0

**Pull request closed without merging**